boskos/boskos.go

```diff
@@ -0,0 +1,383 @@
+
+		if req.Method != "POST" {
+			logrus.Warning("[BadRequest]method %v, expect POST", req.Method)
+			http.Error(res, "/reset only accept POST method", http.StatusBadRequest)
+			return
+		}
+
+		rtype := req.URL.Query().Get("type")
+		state := req.URL.Query().Get("state")
+		expireStr := req.URL.Query().Get("expire")
+		dest := req.URL.Query().Get("dest")
+
+		logrus.Infof("%v, %v, %v, %v", rtype, state, expireStr, dest)
+
+		if rtype == "" || state == "" || expireStr == "" || dest == "" {
+			logrus.Warning("[BadRequest]type, state, expire and dest must present in the request")
+			http.Error(res, "Type, state, expire and dest must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		expire, err := time.ParseDuration(expireStr)
+		if err != nil {
+			logrus.WithError(err).Errorf("Invalid expire: %v", expireStr)
+			http.Error(res, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		rmap := r.Reset(rtype, state, expire, dest)
+		resJSON, err := json.Marshal(rmap)
+		if err != nil {
+			logrus.WithError(err).Errorf("json.Marshal failed : %v", rmap)
+			http.Error(res, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		logrus.Infof("Resource reset: %v", string(resJSON))
+		fmt.Fprint(res, string(resJSON))
+	}
+}
+
+func handleUpdate(r *Ranch) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleUpdate").Infof("From %v", req.RemoteAddr)
+
+		if req.Method != "POST" {
+			logrus.Warning("[BadRequest]method %v, expect POST", req.Method)
+			http.Error(res, "/update only accept POST method", http.StatusBadRequest)
+			return
+		}
+
+		// validate vars
+		name := req.URL.Query().Get("name")
+		if name == "" {
+			logrus.Warning("[BadRequest]name must present in the request")
+			http.Error(res, "Name must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		if err := r.Update(name); err != nil {
+			logrus.WithError(err).Errorf("Update failed: %v", name)
+			http.Error(res, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		logrus.Info("Updated resource %v", name)
+	}
+}
+
+func handleMetric(r *Ranch) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleMetric").Infof("From %v", req.RemoteAddr)
+
+		if req.Method != "GET" {
+			logrus.Warning("[BadRequest]method %v, expect GET", req.Method)
+			http.Error(res, "/metric only accept GET method", http.StatusBadRequest)
+			return
+		}
+
+		fmt.Fprint(res, "To be implemented.\n")
+	}
+}
+
+type Resource struct {
+	Type       string    `json:"type"`
+	Name       string    `json:"name"`
+	State      string    `json:"state"`
+	Owner      string    `json:"owner"`
+	LastUpdate time.Time `json:"lastupdate"`
+}
+
+type Ranch struct {
+	Resources   []*Resource
+	lock        sync.RWMutex
+	storagePath string
+}
+
+func NewRanch(config string, storage string) (*Ranch, error) {
```

Comment:
Seems like initial launch of boskos might be a little awkward... How do I get the initial config into the PD?